src/core/qgsvectorlayerupdater.cpp

```diff
@@ -0,0 +1,87 @@
+#include "qgsvectorlayerupdater.h"
+
+
+QgsVectorLayerUpdater::QgsVectorLayerUpdater(
+  QgsVectorLayer &destinationLayer,
+  const QMap<QString, QgsExpression> fieldsMap,
+  QgsExpressionContext &context,
+  const QList<QString> primaryKeys,
+  const bool bypassEditBuffer )
+  : mLayer( destinationLayer )
+  , mFieldsMap( fieldsMap )
+  , mContext( context )
+  , mPrimaryKeys( primaryKeys )
+  , mBypassEditBuffer( bypassEditBuffer )
+{
+
+}
+
+bool QgsVectorLayerUpdater::addFeature( QgsFeature &feature, QgsFeatureSink::Flags flags )
+{
+  Q_UNUSED( flags );
+
+  mContext.setFeature( ( const QgsFeature & ) feature );
+
+  QgsFeature f = QgsFeature();
```

Comment:
If you're creating a totally new feature to insert into a vector layer, you should use QgsVectorLayerUtils::createFeature so that default value clauses (such as postgres sequences) are correctly handled. This gets a bit tricky here -- you should ONLY call that method for new features. So I think here you may need to refactor this method to move the feature creation/updation to after you've determined whether you're updating an existing feature or not.